tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc

```diff
@@ -120,6 +120,172 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
   return true;
 }
 
+namespace {
+bool Is1dSliceWithoutStrides(const HloInstruction* instr) {
+  return instr->opcode() == HloOpcode::kSlice &&
+         1 == instr->slice_starts().size() &&
+         1 == instr->slice_limits().size() &&
+         1 == instr->slice_strides().size() &&
+         1 == instr->slice_strides().at(0);
+}
+
+bool IsSliceInputFusion(const HloInstruction& unnested_hlo) {
+  if (!unnested_hlo.IsInputFusion()) {
+    return false;
+  }
+  const HloInstruction* root = unnested_hlo.fused_expression_root();
+  if (root->opcode() != HloOpcode::kTuple) {
+    return false;
+  }
+  return absl::c_all_of(root->operands(), [](const HloInstruction* instr) {
+    return Is1dSliceWithoutStrides(instr);
+  });
+}
+
+// Returns whether the concat is used in an elementwise manner.
+// A concat followed by slices is considered (effectively) elementwise if the
+// slices combinedly is a reverse function of the concat.
+// prev_concat_opnd_idx: previously seen concat and its operand index.
+//     prev_concat_opnd_idx.first is nullptr if no previously seen concat.
+bool ConcatIsEffectivelyElementwise(
+    const HloInstruction& concat, const HloInstruction& operand,
+    std::pair<const HloInstruction*, int64>* prev_concat_opnd_idx,
+    const HloInstruction** slice_to_recover_opnd) {
+  // First, check if this concat is in the below pattern. Also, we check
+  // that the slices combinedly are in effect a reverse function of the concat.
+  //
+  //     Concat
+  //     |    |
+  //     v    v
+  //   Slice Slice
+  //
+  std::vector<HloInstruction*> users = concat.users();
+  if (!absl::c_all_of(users, [](const HloInstruction* i) {
+        return Is1dSliceWithoutStrides(i);
+      })) {
+    // Limit our supported cases to 1 dimensional slices.
+    return false;
+  }
+  absl::c_sort(users, [](const HloInstruction* a, const HloInstruction* b) {
+    return a->slice_starts().at(0) < b->slice_starts().at(0);
+  });
+  // Verify that each operand to the concat is reversed by a slice.
+  if (users.size() != concat.operand_count() ||
+      concat.operand_count() != concat.unique_operands().size()) {
+    return false;
+  }
+  int64 prev_limit = 0;
+  for (int64 i = 0; i < users.size(); ++i) {
+    const HloInstruction* u = users[i];
+    int64 slice_size = u->slice_limits().at(0) - u->slice_starts().at(0);
+    if (u->slice_starts().at(0) != prev_limit ||
+        slice_size != ShapeUtil::ElementsIn(concat.operand(i)->shape())) {
+      return false;
+    }
+    prev_limit = u->slice_limits().at(0);
+  }
+
+  // If we have seen other concats, make sure they are identical.
+  int64 operand_idx = concat.operand_index(&operand);
+  *slice_to_recover_opnd = users.at(operand_idx);
+  if (prev_concat_opnd_idx->first == nullptr) {
+    prev_concat_opnd_idx->first = &concat;
+    prev_concat_opnd_idx->second = operand_idx;
+  } else {
+    bool is_concat_identical = prev_concat_opnd_idx->first->Identical(
```

Comment:
Also, why do you need the identical check instead of checking if they are exactly the same object? I think doing the same object check is simpler and cover what we need. If one input feed 2 concat, then it is a case we can't support or do not need.

Reply:
(Unfortunately, ) the same transitive use closure can be fed to multiple concats.

Note that one concat is inserted to each output of the fusion candidates. I added some more descriptions and hopefully it is a bit clearer.